### PR TITLE
ENH: Nearest-neighbor chain algorithm for hierarchical clustering

### PR DESCRIPTION
**Pull request by nmayorov**
Hi, I implemented a new algorithm https://en.wikipedia.org/wiki/Nearest-neighbor_chain_algorithm

Going from O(n^3) to O(n^2) is clearly visible. Note, that it is not applicable to `median` and `centroid` schemas, thus no improvement there. I will investigate if it's possible to achieve speed up for these schemas too. Single linkage was already efficient by SLINK algorithm.

I didn't add any tests, because there is already a very good test suite and the algorithm works as a substitution with the equivalent outcome (which is really convenient).

Below a small benchmarking script. Comments are very welcome.

``` Python
from __future__ import print_function
from time import time
import numpy as np
from scipy.cluster.hierarchy import _LINKAGE_METHODS
from scipy.cluster.hierarchy import linkage as linkage_new
from scipy_.cluster.hierarchy import linkage as linkage_orig  # Version from anaconda
from sklearn.datasets import make_blobs


X, _ = make_blobs(n_samples=3000, centers=100, random_state=0)
print("{:^10} {:^10} {:^10} {:^10}".format(
    'Method', 'Time orig.', 'Time new', 'Match'))
print("-" * 40)
for method in _LINKAGE_METHODS:
    t0 = time()
    Z_orig = linkage_orig(X, method=method)
    t1 = time()
    Z_new = linkage_new(X, method=method)
    t2 = time()
    print("{:^10} {:^10.3f} {:^10.3f} {:^10}".format(
        method, t1 - t0, t2 - t1, np.allclose(Z_orig, Z_new)))
```

Output:

```
  Method   Time orig.  Time new    Match   
----------------------------------------
  median     10.966     9.784        1     
   ward      10.802     0.267        1     
  single     0.094      0.096        1     
 centroid    11.501     10.215       1     
 complete    10.996     0.252        1     
 average     11.557     0.264        1     
 weighted    11.163     0.238        1     
[Finished in 88.8s]
```

### CODE REVIEW

**Comment by rgommers** [comment]
Nice. @richardtsai you may be interested in this

**Comment by rgommers** [comment]
The benchmark above could easily be converted into an ASV benchmark, that would be nice. The only issue is the ``make_blobs`` function, maybe can be replaced by some other data generator that doesn't add a dependency?

File: scipy/cluster/hierarchy.py [hunk 1]
```diff
@@ -619,55 +618,53 @@ def linkage(y, method='single', metric='euclidean'):
     Z : ndarray
         The hierarchical clustering encoded as a linkage matrix.
 
+    Notes
+    -----
+    1. For method 'single' an optimized algorithm called SLINK is implemented,
+       which has :math:`O(n^2)` time complexity and :math:`O(n)` memory
```
Comment (richardtsai) [inline scipy/cluster/hierarchy.py]: The space complexity is in fact O(n^2) in our current implementation, as we pass a `dist` array which has n(n-1)/2 elements, into the linkage function. An O(n) space implementation of SLINK requires computing distances on the fly, which seems not to fit well into our API design.

File: scipy/cluster/_hierarchy.pyx [hunk 2]
```diff
@@ -783,6 +789,141 @@ def linkage(double[:] dists, double[:, :] Z, int n, int method):
...
+    cdef int next_label
+
+    def __init__(self, int n):
+        self.parent = np.arange(2 * n - 1, dtype=np.intc)
+        self.next_label = n
+
+    cdef merge(self, int x, int y):
+        self.parent[x] = self.next_label
+        self.parent[y] = self.next_label
+        self.next_label += 1
```
Comment (richardtsai) [inline scipy/cluster/_hierarchy.pyx]: Does it work if we use merge-by-rank and avoid creating new nodes here?

**Comment by richardtsai** [comment]
Awesome work @nmayorov ! Now we finally have an implementation comparable with [fastcluster](http://danifold.net/fastcluster.html?section=1)!

File: scipy/cluster/hierarchy.py (hunk 1 above)
Comment (nmayorov) [inline scipy/cluster/hierarchy.py]: Right, I knew this, I was implying "additional memory". Is it better to name it "additional memory" or maybe say that all algorithms have O(n^2) for memory in our implementation?

**Comment by nmayorov** [comment]
It was a paper by the author of "fastcluster" which helped me to implement the correct and full version of NN-chain (wiki version missing the part of restoring the standard order of merges).

File: scipy/cluster/_hierarchy.pyx (hunk 2 above)
Comment (nmayorov) [inline scipy/cluster/_hierarchy.pyx]: Are you referring to what is called as "union by rank" in [wikipedia](https://en.wikipedia.org/wiki/Disjoint-set_data_structure)? From the first and conceptual look: using ranks tends to create balanced trees, but in our clustering problem the tree structure is generally arbitrary, so it seems like some contradiction.

Do you see any problems with this approach? (Basically I took it from the paper and it looked very logical and straightforward for me).

**Comment by nmayorov** [comment]
@rgommers I added a very simple benchmark. Is it good enough, you think?

As for data to cluster, I used a random matrix: for benchmarking purpose it doesn't make any difference, so I decided to use this boring and simple `randn`.

**Comment by nmayorov** [comment]
Well, if nobody has any objections, I think it is good enough to be merged.

@rgommers what do you think?

**Comment by rgommers** [comment]
> @rgommers I added a very simple benchmark. Is it good enough, you think?

This looks fine to me. Almost all ``asv`` benchmarks look very simple, but it's more that ``asv`` is very nice than that the benchmarks are too simplistic:)

**Comment by rgommers** [comment]
It looks good to me, passes all tests, has a benchmark and everyone seems happy - so in it goes. Thanks Nikolay!

**Pull request merged**